generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -0,0 +1,117 @@
+import java.util.regex.Pattern;
+
+import static com.google.devtools.build.workspace.maven.Rule.MAVEN_CENTRAL_URL;
+
+public class VersionResolver {
+
+  private final VersionHTMLScraper downloader;
+
+  VersionResolver() {
+    this(new VersionHTMLScraper(MAVEN_CENTRAL_URL));
+  }
+
+  @VisibleForTesting
+  VersionResolver(VersionHTMLScraper downloader) {
+    this.downloader = downloader;
+  }
+
+  /**
+   * Takes a version specification (as defined in
+   * http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html) and scrapes the maven
+   * remote repository for the highest available version that is accepted by the provided version
+   * specification. For example, 1.2.3 should be interpreted as 1.2.3+. If the available versions are
+   * {1.2.1, 1.2.3, and 1.2.4}, then this will return 1.2.4.
+   */
+  String resolveVersion(String groupId, String artifactId, String versionSpec)
+      throws Resolver.InvalidArtifactCoordinateException {
+
+    List<String> potentialVersions;
+    try {
+      potentialVersions = getPotentialVersions(groupId, artifactId, versionSpec);
+
+    } catch (IOException e) {
+      throw new Resolver.InvalidArtifactCoordinateException(
+          messageForInvalidArtifact(groupId, artifactId, versionSpec));
+    }
+    return Iterables.getLast(potentialVersions);
+  }
+
+  /**
+   * Scrapes maven page for an artifact for all available versions, and returns a list of versions
+   * that meet the requirements provided by the version specifications.
+   *
+   * One important property of maven artifact page is that all versions are sorted from earliest to latest.
+   */
+  // TODO(petros): add support for repositories other than Maven central.
+  private List<String> getPotentialVersions(String groupId, String artifactId, String versionSpec)
+      throws IOException, Resolver.InvalidArtifactCoordinateException {
+    List<String> hyperLinks = downloader.scrapeHyperLinksForArtifact(groupId, artifactId);
+
+    List<String> potentialVersions = new ArrayList<>();
+    for (String link : hyperLinks) {
+      String potentialVersion = removeForwardSlashAtEnd(link);
+      if (isSnapshotVersion(potentialVersion)
+          || !containedInSpecifiedRange(groupId, artifactId, versionSpec, potentialVersion)) {
+        continue;
+      }
+      potentialVersions.add(potentialVersion);
+    }
+    if (potentialVersions.isEmpty()) {
+      throw new Resolver.InvalidArtifactCoordinateException(
+          messageForInvalidArtifact(groupId, artifactId, versionSpec));
+    }
+    return potentialVersions;
+  }
+
+  /** Checks if a given version is within the bounds of a version range */
+  private static boolean containedInSpecifiedRange(
+      String groupId, String artifactId, String versionSpec, String potentialVersion)
+      throws Resolver.InvalidArtifactCoordinateException {
+
+    if (versionSpec.isEmpty()) {
+      return true;
+    }
+    versionSpec = transformBasicVersionIntoRange(versionSpec);
```

Comment:
You shouldn't need this.  VersionRange.createFromVersionSpec("1.2.3").containsVersion("1.2.4") is true.